src/heart_disease_predictor_report.ipynb

```diff
@@ -240,7 +240,94 @@
    "cell_type": "code",
-   "execution_count": 4,
+   "execution_count": 30,
+   "id": "1fa40873-e180-4127-873d-849c7c93bad5",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Column 'age' has acceptable missingness (0.00%).\n",
+      "Column 'sex' has acceptable missingness (0.00%).\n",
+      "Column 'cp' has acceptable missingness (0.00%).\n",
+      "Column 'trestbps' has acceptable missingness (0.00%).\n",
+      "Column 'chol' has acceptable missingness (0.00%).\n",
+      "Column 'fbs' has acceptable missingness (0.00%).\n",
+      "Column 'restecg' has acceptable missingness (0.00%).\n",
+      "Column 'thalach' has acceptable missingness (0.00%).\n",
+      "Column 'exang' has acceptable missingness (0.00%).\n",
+      "Column 'oldpeak' has acceptable missingness (0.00%).\n",
+      "Column 'slope' has acceptable missingness (0.00%).\n",
+      "Column 'ca' has acceptable missingness (1.32%).\n",
+      "Column 'thal' has acceptable missingness (0.66%).\n",
+      "Column 'num' has acceptable missingness (0.00%).\n",
+      "------------------------\n",
+      "Data type validation passed successfully.\n",
+      "------------------------\n",
+      "No duplicate rows found in the dataset.\n"
+     ]
+    }
+   ],
+   "source": [
+    "import pandera as pa\n",
+    "from pandera import Column, Check, DataFrameSchema\n",
+    "# ---- Check 4: Missingness not beyond expected threshold ----\n",
+    "# Set acceptable threshold for missing data (e.g., no more than 10%)\n",
+    "acceptable_threshold = 0.1\n",
+    "\n",
+    "# Calculate the proportion of missing values in each column\n",
+    "missing_proportions = heart_disease.isna().mean()\n",
+    "\n",
+    "# Check if any column exceeds the acceptable threshold\n",
+    "for column, proportion in missing_proportions.items():\n",
+    "    if proportion > acceptable_threshold:\n",
+    "        print(f\"Warning: Missing values in column '{column}' exceed the acceptable threshold ({proportion:.2%}).\")\n",
+    "    else:\n",
+    "        print(f\"Column '{column}' has acceptable missingness ({proportion:.2%}).\")\n",
+    "print(\"------------------------\")\n",
+    "\n",
+    "# ---- Check 5: Correct data types in each column ----\n",
+    "heart_disease[\"ca\"] = heart_disease[\"ca\"].astype('float64')\n",
+    "# Define the schema\n",
+    "schema = DataFrameSchema({\n",
+    "    \"age\": Column(float),\n",
+    "    \"sex\": Column(float),\n",
+    "    \"cp\": Column(float),\n",
+    "    \"trestbps\": Column(float),\n",
+    "    \"chol\": Column(float),\n",
+    "    \"fbs\": Column(float),\n",
+    "    \"restecg\": Column(float),\n",
+    "    \"thalach\": Column(float),\n",
+    "    \"exang\": Column(float),\n",
+    "    \"oldpeak\": Column(float),\n",
+    "    \"slope\": Column(float),\n",
+    "    \"ca\": Column(float, nullable=True),\n",
+    "    \"thal\": Column(str, nullable=True),\n",
```

Comment:
curious why thal left as object when ca was converted, I guess either could be either though and this isn't an issue blocking merge

Reply:
what i was thinking is that 'thal' is left as an object because it represents categorical data rather than a numeric type, whereas 'ca' was converted to float since it should represent a numerical value with possible missing data.
correct me if its wrong

Reply:
ohh ok, yeah that makes sense to me now, thanks for explaining